src/main/java/bio/terra/service/filedata/azure/tables/TableDirectoryDao.java

```diff
@@ -0,0 +1,241 @@
+import com.azure.data.tables.TableClient;
+import com.azure.data.tables.TableServiceClient;
+import com.azure.data.tables.models.ListEntitiesOptions;
+import com.azure.data.tables.models.TableEntity;
+import com.azure.data.tables.models.TableServiceException;
+import com.azure.data.tables.models.TableTransactionAction;
+import com.azure.data.tables.models.TableTransactionActionType;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+/**
+ * Paths and document names Azure uses forward slash (/) for its path separator. We also use forward
+ * slash in our file system paths. To get uniqueness of files, we want to name files with their full
+ * path. Otherwise, two threads could create the same file as two different documents. That would
+ * not do at all.
+ *
+ * <p>We solve this problem by using Azure document names that replace the forward slash in our
+ * paths with character 0x1c - the unicode file separator character. That allows documents to be
+ * named with their full names. (See https://www.compart.com/en/unicode/U+001C) That replacement is
+ * <strong>only</strong> used for the Azure document names. All the other paths we process use the
+ * forward slash separator.
+ *
+ * <p>We need a root directory to hold the other directories. Since we are doing Azure lookup by
+ * document name, the root directory needs a name. We call it "_dr_"; it could be anything, but it
+ * helps when viewing Azure in the console that it has an obvious name.
+ *
+ * <p>We don't store the root directory in the paths stored in file and directory entries. It is
+ * only used for the Azure lookup. When we refer to paths in the code we talk about: - lookup path -
+ * the path used for the Azure lookup. When building this path (and only this path) we prepended it
+ * with "_dr_" as the name for the root directory. - directory path - the directory path to the
+ * directory containing entry - not including the entry name - full path - the full path to the
+ * entry including the entry name.
+ *
+ * <p>Within the document we store the directory path to the entry and the entry name. That lets us
+ * use the indexes to find the entries in a directory.
+ *
+ * <p>It is an invariant that there are no empty directories. When a directory becomes empty on a
+ * delete, it is deleted. When a directory is needed, we create it. That is all done within
+ * transactions so there is never a time where the externally visible state violates that invariant.
+ */
+@Component
+public class TableDirectoryDao {
+  private final Logger logger = LoggerFactory.getLogger(FireStoreDirectoryDao.class);
+  private static final String TABLE_NAME = "dataset";
+  private static final String PARTITION_KEY = "partitionKey";
```

Comment:
Is there a domain/functional value that can be used as a partition key - e.g., filesystem name,  root directory, parent directory, etc.? I understand the tendency towards simplicity and therefore start with a fixed partition.  But it is always better to think through this problem at the beginning vs. later and dealing with unpartitioned data.

Reply:
That's a good point! Each dataset/snapshot will have a separate storage account with its own tables, so partitioning this table by dataset id won't make a difference. I think using the parent directory would work since we can derive that from the file path when writing/retrieving rows.